### ISSUE TITLE
pageserver management API doesn't poll request handlers to completion if client hangs up [P:1] [S:1]

### ISSUE THREAD

**Issue by problame** [neondatabase/neon#3478]
## Steps to reproduce

1. Modify a management API endpoint to `tokio::time::sleep().await` for 10 seconds, and log something before and after.
2. Make a request to that endpoint using `curl`.
3. Ctrl-C `curl`

## Expected result

Log output:

```
T+00 the message you logged before sleep
T+10 the message you logged after sleep
```

## Actual result

```
T+00 the message you logged before sleep
```

## Environment

All

---

## Why Is This A Problem?

Some management API handlers are not cancel-safe. In fact, I suspect most of them aren't.
One example is `/detach`.
If the client hangs up _after_ we've transitioned the tenant to `Stopping`, but before `mgr::detach_tenant` returns, the following happens:
- the detach operation isn't finished
- the tenant remains in `Stopping` state
- subsequent detatch operations fail because we fail them if the tenant is `Stopping` (rightly so)
- after pageserver restart, the tenant will come up as attached again
  - In the worst case, this can result in **split brain**. Right now this is not an issue, but once we add automatic tenant relocation, it will be (cc @LizardWizzard )

## Solution

Spawn all management API tasks as `task_mgr` task so that they run to completion, even if the webserver stops polling the futures.

The implementation should create some kind of middleware to do this, and that middleware should also pass in a `RequestContext` to the task. Once `RequestContext` learns cancellation, the middleware should cancel the `RequestContext` when the request handling future (not the task that is spawned!) is dropped.

If this issue is worked on before `RequestContext` supports cancellation, use a `CancellationToken` instead.

Consider reducing control-plane timeout once issue is fixed: https://github.com/neondatabase/cloud/commit/ff390331f9bdb80d7120d30f8d6d1ea4ee588f37. Maybe some other tweaks will be needed too

**Comment by LizardWizzard** [comment]
Hmm, yeah, the whole thing doesnt look right to me. This is definitely a problem. 

There are several problems with detach. Initially I thought about it in a way that it shouldnt block for a long time, and should be resumable. I e when it fails on something it is safe to retry the request and it will retry the thing that failed, no the operation from the beginning. So it may make sense to make it cancel safe. Detach also should work correctly through power cycle. (This should already be solved with mark file)

From the other side we can go to completely async model, let detach return 202, and let the client poll it till the completion. It will make error reporting harder. Because we need to keep failed operations in memory to give them back during poll phase. Client will have to handle other states, e g if pageserver was restarted when detach was in progress.

**Comment by LizardWizzard** [comment]
Thoughts: See how this relates to changing http framework, think about using spawn for each connection handler. Beware of potential concurrency issues, e g when retry happens when previous call didnt finish

**Comment by koivunej** [comment]
The topic is wrong, it's a feature not a bug that this happens, because there's no way to do the other way around if you'd have different defaults.

I thought we had already spawned a task for detach for example, but we don't, this is causing some issues in prod: https://neondb.slack.com/archives/C03H1K0PGKH/p1680849458972369 -- at least for one tenant upload task shutdown took so long.

**Comment by ololobus** [comment]
> Thoughts: See how this relates to changing http framework, think about using spawn for each connection handler.

If I got it correctly, then if you spawn a new thread and respond with 200 OK to the caller. There is still a chance that pageserver will restart until spawned thread finishes its job, but the caller (control-plane) won't retry call in this case anymore as it got OK already

Another point to consider: in the handler only persistently set some flag that detach requested and process in in the background. Then caller will need to re-check that tenant was actually fully detached

**Comment by LizardWizzard** [comment]
> There is still a chance that pageserver will restart until spawned thread finishes its job, but the caller (control-plane) won't retry call in this case anymore as it got OK already

Obviously it should return accepted or something like that so it would be polled again.

The original detach idea is that it should be cancel safe, i e the retry request should be able to continue half finished operation.

**Comment by problame** [comment]
We can run the detach in a task_mgr task, yet still make the request blocking, e.g., by communicating the detach result to the request handler using a `oneshot`. 

We could also do this systematically in the `RequestSpan` type.

This doesn't _solve_ the underlying issue of detach being not cancel & crash-safe.
But, in practice, it will make the issue much less likely to occur.
So, it might be a good interim mitigation.

In today's pageserver meeting we decided that @LizardWizzard scope out what it takes to make detach cancel & crash-safe. Depending on what comes out of that, we should act on my proposal in the previous paragraph.
(Do we have an issue for that/)

In general, I think our code is not written with cancel-by-drop-safety in mind, so, the default should be that request handlers are run to completion in a task, even if the client hangs up.
Request handlers should check for cancellation (mechanism TBD) so that they complete faster if the client hangs up.
That's what this issue was supposed to be about, and I don't get why @koivunej thinks it's ill-named.

**Comment by hlinnaka** [comment]
This was fixed in PR #4314.
